### PR TITLE
Synthetic data generation

### PR DESCRIPTION
**Pull request by mateusz-wosinski-ds**
Description: this MR implements generation of synthetic data with given keywords

### CODE REVIEW

File: libs/langchain/langchain/chains/synthetic_data/__init__.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+def create_data_generation_chain(
+    llm: BaseLanguageModel,
+    prompt: Optional[PromptTemplate] = None,
+) -> Chain:
+    """Creates a chain that generates synthetic sentences with
+     provided fields.
+
+    Args:
+        llm: The language model to use.
+        verbose: Whether to print the prompt or not.
```
Comment (eryk-dsai) [inline libs/langchain/langchain/chains/synthetic_data/__init__.py]: I think there is a mismatch between docstring and function arguments

File: libs/langchain/langchain/chains/synthetic_data/prompts.py [hunk 2]
```diff
@@ -0,0 +1,10 @@
+from langchain.prompts.prompt import PromptTemplate
+
+sentence_template = """Given the following fields, create a sentence about them. Make it detailed and interesting.
+
+Fields:
+{fields}
+Sentence:
```
Comment (maks-operlejn-ds) [inline libs/langchain/langchain/chains/synthetic_data/prompts.py]: Apparently newline at the end of the prompt isn't a good idea - GPT probably won't care, but open-source models (GPT-J) tends to screw up the probabilities :smile: 

I encountered this while testing GPT-J:
image

In LangChain, from what I've looked at, they don't give that newline character either

File: libs/langchain/langchain/chains/synthetic_data/prompts.py [hunk 3]
```diff
@@ -0,0 +1,9 @@
+from langchain.prompts.prompt import PromptTemplate
+
+sentence_template = """Given the following fields, create a sentence about them. Make it detailed and interesting.
```
Comment (piotr-grodek-dsai) [inline libs/langchain/langchain/chains/synthetic_data/prompts.py]: `Use every field.` - I would likely add it as well.

File: libs/langchain/langchain/chains/synthetic_data/__init__.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
+from typing import Any, Dict, List, Optional
+
+from langchain.chains.base import Chain
+from langchain.chains.llm import LLMChain
+from langchain.chains.synthetic_data.prompts import SENTENCE_PROMPT
+from langchain.prompts import PromptTemplate
```
Comment (eryk-dsai) [inline libs/langchain/langchain/chains/synthetic_data/__init__.py]: @piotr-grodek-dsai @maks-operlejn-ds @mateusz-wosinski-ds 
if `PromptTemplate` and `BaseLanguageModel` are imported only for type hints, is it worth to import `TYPE_CHECKING` from `typing` and add 
```
if TYPE_CHECKING:
    ...
```
logic here ?

I've encountered this suggestion from LangChain team during my PR, so maybe it's good practice that I didn't know

**Review by eryk-dsai — APPROVED** [review]

File: libs/langchain/langchain/chains/synthetic_data/__init__.py (hunk 4 above)
Comment (piotr-grodek-dsai) [reply libs/langchain/langchain/chains/synthetic_data/__init__.py]: Yes, if something is required only for typing then it should be imported under if.

**Pull request closed without merging**